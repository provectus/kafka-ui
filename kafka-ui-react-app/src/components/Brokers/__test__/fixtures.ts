import { BrokersLogdirs } from 'generated-sources';

export const brokersPayload = [
  { id: 1, host: 'b-1.test.kafka.amazonaws.com', port: 9092 },
  { id: 2, host: 'b-2.test.kafka.amazonaws.com', port: 9092 },
];

export const clusterStatsPayload = {
  brokerCount: 2,
  activeControllers: 1,
  onlinePartitionCount: 138,
  offlinePartitionCount: 0,
  inSyncReplicasCount: 239,
  outOfSyncReplicasCount: 0,
  underReplicatedPartitionCount: 0,
  diskUsage: [
    { brokerId: 0, segmentSize: 334567, segmentCount: 245 },
    { brokerId: 1, segmentSize: 12345678, segmentCount: 121 },
  ],
  version: '2.2.1',
};

const partition = {
  broker: 2,
  offsetLag: 0,
  partition: 2,
  size: 0,
};
const topics = {
  name: '_confluent-ksql-devquery_CTAS_NUMBER_OF_TESTS_59-Aggregate-Aggregate-Materialize-changelog',
  partitions: [partition],
};

export const brokerLogDirsPayload: BrokersLogdirs[] = [
  {
    error: 'NONE',
    name: '/opt/kafka/data-0/logs',
    topics: [
      {
        ...topics,
        partitions: [partition, partition, partition],
      },
      topics,
      {
        ...topics,
        partitions: [],
      },
    ],
  },
];
