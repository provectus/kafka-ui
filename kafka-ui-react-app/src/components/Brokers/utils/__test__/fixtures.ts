import { BrokerMetrics } from 'generated-sources';

export const brokerMetricsPayload: BrokerMetrics = {
  segmentSize: 23,
  segmentCount: 23,
  metrics: [
    {
      name: 'TotalFetchRequestsPerSec',
      labels: {
        canonicalName:
          'kafka.server:name=TotalFetchRequestsPerSec,topic=_connect_status,type=BrokerTopicMetrics',
      },
      value: 10,
    },
    {
      name: 'ZooKeeperRequestLatencyMs',
      value: 11,
    },
    {
      name: 'RequestHandlerAvgIdlePercent',
    },
  ],
};
export const transformedBrokerMetricsPayload =
  '{"segmentSize":23,"segmentCount":23,"metrics":[{"name":"TotalFetchRequestsPerSec","labels":{"canonicalName":"kafka.server:name=TotalFetchRequestsPerSec,topic=_connect_status,type=BrokerTopicMetrics"},"value":10},{"name":"ZooKeeperRequestLatencyMs","value":11},{"name":"RequestHandlerAvgIdlePercent"}]}';
