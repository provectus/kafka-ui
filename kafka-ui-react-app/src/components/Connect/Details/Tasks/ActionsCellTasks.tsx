import React from 'react';
import { Task } from 'generated-sources';
import { CellContext } from '@tanstack/react-table';
import useAppParams from 'lib/hooks/useAppParams';
import { useRestartConnectorTask } from 'lib/hooks/api/kafkaConnect';
import { Dropdown, DropdownItem } from 'components/common/Dropdown';
import { RouterParamsClusterConnectConnector } from 'lib/paths';

const ActionsCellTasks: React.FC<CellContext<Task, unknown>> = ({ row }) => {
  const { id } = row.original;
  const routerProps = useAppParams<RouterParamsClusterConnectConnector>();
  const restartMutation = useRestartConnectorTask(routerProps);

  const restartTaskHandler = (taskId?: number) => {
    if (taskId === undefined) return;
    restartMutation.mutateAsync(taskId);
  };

  return (
    <Dropdown>
      <DropdownItem
        onClick={() => restartTaskHandler(id?.task)}
        danger
        confirm="Are you sure you want to restart the task?"
      >
        <span>Restart task</span>
      </DropdownItem>
    </Dropdown>
  );
};

export default ActionsCellTasks;
