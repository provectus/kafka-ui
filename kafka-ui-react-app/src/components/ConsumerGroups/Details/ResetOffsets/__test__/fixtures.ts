export const expectedOutputs = {
  EARLIEST: {
    partitions: [0],
    partitionsOffsets: [
      {
        offset: undefined,
        partition: 0,
      },
    ],
    resetType: 'EARLIEST',
    topic: '__amazon_msk_canary',
  },
  LATEST: {
    partitions: [0],
    partitionsOffsets: [
      {
        offset: undefined,
        partition: 0,
      },
    ],
    resetType: 'LATEST',
    topic: '__amazon_msk_canary',
  },
  OFFSET: {
    partitions: [0],
    partitionsOffsets: [
      {
        offset: '10',
        partition: 0,
      },
    ],
    resetType: 'OFFSET',
    topic: '__amazon_msk_canary',
  },
};
