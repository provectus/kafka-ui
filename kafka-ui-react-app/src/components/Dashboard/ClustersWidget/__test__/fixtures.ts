import { Cluster, ServerStatus } from 'generated-sources';

export const onlineCluster: Cluster = {
  name: 'secondLocal',
  defaultCluster: false,
  status: ServerStatus.Online,
  brokerCount: 1,
  onlinePartitionCount: 6,
  topicCount: 3,
  bytesInPerSec: 0.000030618196853764715,
  bytesOutPerSec: 5.737800890036267075817,
};

export const offlineCluster: Cluster = {
  name: 'local',
  defaultCluster: true,
  status: ServerStatus.Offline,
  brokerCount: 1,
  onlinePartitionCount: 2,
  topicCount: 2,
  bytesInPerSec: 8000.0000067376808542600021,
  bytesOutPerSec: 0.8153063567297119490871,
};

export const clusters: Cluster[] = [onlineCluster, offlineCluster];
