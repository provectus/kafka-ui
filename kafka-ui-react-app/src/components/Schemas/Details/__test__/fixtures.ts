import { SchemaSubject, SchemaType } from 'generated-sources';
import {
  schemaVersion1,
  schemaVersion2,
  schemaVersionWithNonAsciiChars,
} from 'redux/reducers/schemas/__test__/fixtures';

export const versionPayload = [
  schemaVersion1,
  schemaVersion2,
  schemaVersionWithNonAsciiChars,
];
export const versionEmptyPayload = [];

export const jsonSchema: SchemaSubject = {
  subject: 'test',
  version: '15',
  id: 1,
  schema:
    '{"type":"record","name":"MyRecord1","namespace":"com.mycompany","fields":[{"name":"id","type":"long"}]}',
  compatibilityLevel: 'BACKWARD',
  schemaType: SchemaType.JSON,
};

export const protoSchema: SchemaSubject = {
  subject: 'test_proto',
  version: '1',
  id: 2,
  schema:
    'syntax = "proto3";\npackage com.indeed;\n\nmessage MyRecord {\n  int32 id = 1;\n  string name = 2;\n}\n',
  compatibilityLevel: 'BACKWARD',
  schemaType: SchemaType.PROTOBUF,
};
