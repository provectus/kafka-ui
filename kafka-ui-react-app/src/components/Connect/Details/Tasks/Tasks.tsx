import React from 'react';
import { useConnectorTasks } from 'lib/hooks/api/kafkaConnect';
import useAppParams from 'lib/hooks/useAppParams';
import { RouterParamsClusterConnectConnector } from 'lib/paths';
import { ColumnDef, Row } from '@tanstack/react-table';
import { Task } from 'generated-sources';
import Table, { TagCell } from 'components/common/NewTable';

import ActionsCellTasks from './ActionsCellTasks';

const ExpandedTaskRow: React.FC<{ row: Row<Task> }> = ({ row }) => {
  return <div>{row.original.status.trace}</div>;
};

const MAX_LENGTH = 100;

const Tasks: React.FC = () => {
  const routerProps = useAppParams<RouterParamsClusterConnectConnector>();
  const { data = [] } = useConnectorTasks(routerProps);

  const columns = React.useMemo<ColumnDef<Task>[]>(
    () => [
      { header: 'ID', accessorKey: 'status.id' },
      { header: 'Worker', accessorKey: 'status.workerId' },
      { header: 'State', accessorKey: 'status.state', cell: TagCell },
      {
        header: 'Trace',
        accessorKey: 'status.trace',
        enableSorting: false,
        cell: ({ getValue }) => {
          const trace = getValue<string>() || '';
          return trace.toString().length > MAX_LENGTH
            ? `${trace.toString().substring(0, MAX_LENGTH - 3)}...`
            : trace;
        },
        meta: { width: '70%' },
      },
      {
        id: 'actions',
        header: '',
        cell: ActionsCellTasks,
      },
    ],
    []
  );

  return (
    <Table
      columns={columns}
      data={data}
      emptyMessage="No tasks found"
      enableSorting
      getRowCanExpand={(row) => row.original.status.trace?.length > 0}
      renderSubComponent={ExpandedTaskRow}
    />
  );
};

export default Tasks;
