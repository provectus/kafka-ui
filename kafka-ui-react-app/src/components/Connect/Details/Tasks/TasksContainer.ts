import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { RootState } from 'redux/interfaces';
import { fetchConnectorTasks } from 'redux/reducers/connect/connectSlice';
import {
  getAreConnectorTasksFetching,
  getConnectorTasks,
} from 'redux/reducers/connect/selectors';

import Tasks from './Tasks';

const mapStateToProps = (state: RootState) => ({
  areTasksFetching: getAreConnectorTasksFetching(state),
  tasks: getConnectorTasks(state),
});

const mapDispatchToProps = {
  fetchTasks: fetchConnectorTasks,
};

export default withRouter(connect(mapStateToProps, mapDispatchToProps)(Tasks));
