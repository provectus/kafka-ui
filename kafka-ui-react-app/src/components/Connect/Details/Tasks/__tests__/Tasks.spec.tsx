import React from 'react';
import { render, WithRoute } from 'lib/testHelpers';
import { clusterConnectConnectorTasksPath } from 'lib/paths';
import Tasks from 'components/Connect/Details/Tasks/Tasks';
import { tasks } from 'lib/fixtures/kafkaConnect';
import { screen, within, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import {
  useConnectorTasks,
  useRestartConnectorTask,
} from 'lib/hooks/api/kafkaConnect';
import { Task } from 'generated-sources';

jest.mock('lib/hooks/api/kafkaConnect', () => ({
  useConnectorTasks: jest.fn(),
  useRestartConnectorTask: jest.fn(),
}));

const path = clusterConnectConnectorTasksPath('local', 'ghp', '1');

const restartConnectorMock = jest.fn();

describe('Tasks', () => {
  beforeEach(() => {
    (useRestartConnectorTask as jest.Mock).mockImplementation(() => ({
      mutateAsync: restartConnectorMock,
    }));
  });

  const renderComponent = (currentData: Task[] | undefined = undefined) => {
    (useConnectorTasks as jest.Mock).mockImplementation(() => ({
      data: currentData,
    }));

    render(
      <WithRoute path={clusterConnectConnectorTasksPath()}>
        <Tasks />
      </WithRoute>,
      { initialEntries: [path] }
    );
  };

  it('renders empty table', () => {
    renderComponent();
    expect(screen.getByRole('table')).toBeInTheDocument();
    expect(screen.getByText('No tasks found')).toBeInTheDocument();
  });

  it('renders tasks table', () => {
    renderComponent(tasks);
    expect(screen.getAllByRole('row').length).toEqual(tasks.length + 1);

    expect(
      screen.getByRole('row', {
        name: '1 kafka-connect0:8083 RUNNING',
      })
    ).toBeInTheDocument();
  });

  it('renders truncates long trace and expands', async () => {
    renderComponent(tasks);

    const trace = tasks[2]?.status?.trace || '';
    const truncatedTrace = trace.toString().substring(0, 100 - 3);

    const thirdRow = screen.getByRole('row', {
      name: `3 kafka-connect0:8083 RUNNING ${truncatedTrace}...`,
    });
    expect(thirdRow).toBeInTheDocument();

    const expandedDetails = screen.queryByText(trace);
    //  Full trace is not visible
    expect(expandedDetails).not.toBeInTheDocument();

    await userEvent.click(thirdRow);

    expect(
      screen.getByRole('row', {
        name: trace,
      })
    ).toBeInTheDocument();
  });

  describe('Action button', () => {
    const expectDropdownExists = async () => {
      const firstTaskRow = screen.getByRole('row', {
        name: '1 kafka-connect0:8083 RUNNING',
      });
      expect(firstTaskRow).toBeInTheDocument();
      const extBtn = within(firstTaskRow).getByRole('button', {
        name: 'Dropdown Toggle',
      });
      expect(extBtn).toBeEnabled();
      await userEvent.click(extBtn);
      expect(screen.getByRole('menu')).toBeInTheDocument();
    };

    it('renders action button', async () => {
      renderComponent(tasks);
      await expectDropdownExists();
      expect(
        screen.getAllByRole('button', { name: 'Dropdown Toggle' }).length
      ).toEqual(tasks.length);
      // Action buttons are enabled
      const actionBtn = screen.getAllByRole('menuitem');
      expect(actionBtn[0]).toHaveTextContent('Restart task');
    });

    it('works as expected', async () => {
      renderComponent(tasks);
      await expectDropdownExists();
      const actionBtn = screen.getAllByRole('menuitem');
      expect(actionBtn[0]).toHaveTextContent('Restart task');

      await userEvent.click(actionBtn[0]);
      expect(
        screen.getByText('Are you sure you want to restart the task?')
      ).toBeInTheDocument();

      expect(screen.getByText('Confirm the action')).toBeInTheDocument();
      userEvent.click(screen.getByRole('button', { name: 'Confirm' }));

      await waitFor(() => expect(restartConnectorMock).toHaveBeenCalled());
    });
  });
});
