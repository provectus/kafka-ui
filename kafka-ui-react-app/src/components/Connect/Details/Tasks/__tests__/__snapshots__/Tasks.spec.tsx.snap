// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks view matches snapshot 1`] = `
.c0 {
  width: 100%;
}

.c0 td {
  border-top: 1px #f1f2f3 solid;
  font-size: 14px;
  font-weight: 400;
  padding: 8px 8px 8px 24px;
  color: #171A1C;
  vertical-align: middle;
}

.c0 tbody > tr:hover {
  background-color: #F1F2F3;
}

.c2 {
  font-family: Inter,sans-serif;
  font-size: 12px;
  font-style: normal;
  font-weight: 400;
  line-height: 16px;
  -webkit-letter-spacing: 0em;
  -moz-letter-spacing: 0em;
  -ms-letter-spacing: 0em;
  letter-spacing: 0em;
  text-align: left;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFFFFF;
  color: #73848C;
  cursor: default;
}

.c1 {
  padding: 4px 0 4px 24px;
  border-bottom-width: 1px;
  vertical-align: middle;
}

<table
  className="c0"
>
  <thead>
    <tr>
      <th
        className="c1"
      >
        <span
          className="c2"
        >
          ID
        </span>
      </th>
      <th
        className="c1"
      >
        <span
          className="c2"
        >
          Worker
        </span>
      </th>
      <th
        className="c1"
      >
        <span
          className="c2"
        >
          State
        </span>
      </th>
      <th
        className="c1"
      >
        <span
          className="c2"
        >
          Trace
        </span>
      </th>
      <th
        className="c1"
      >
        <span
          className="c2"
        />
      </th>
    </tr>
  </thead>
  <tbody>
    <tr
      task={
        Object {
          "config": Object {
            "batch.size": "2000",
            "file": "/some/file",
            "task.class": "org.apache.kafka.connect.file.FileStreamSourceTask",
            "topic": "test-topic",
          },
          "id": Object {
            "connector": "first",
            "task": 1,
          },
          "status": Object {
            "id": 1,
            "state": "RUNNING",
            "workerId": "kafka-connect0:8083",
          },
        }
      }
    />
    <tr
      task={
        Object {
          "config": Object {
            "batch.size": "1000",
            "file": "/some/file2",
            "task.class": "org.apache.kafka.connect.file.FileStreamSourceTask",
            "topic": "test-topic",
          },
          "id": Object {
            "connector": "first",
            "task": 2,
          },
          "status": Object {
            "id": 2,
            "state": "FAILED",
            "trace": "Failure 1",
            "workerId": "kafka-connect0:8083",
          },
        }
      }
    />
    <tr
      task={
        Object {
          "config": Object {
            "batch.size": "3000",
            "file": "/some/file3",
            "task.class": "org.apache.kafka.connect.file.FileStreamSourceTask",
            "topic": "test-topic",
          },
          "id": Object {
            "connector": "first",
            "task": 3,
          },
          "status": Object {
            "id": 3,
            "state": "RUNNING",
            "workerId": "kafka-connect0:8083",
          },
        }
      }
    />
  </tbody>
</table>
`;

exports[`Tasks view matches snapshot when fetching tasks 1`] = `<mock-PageLoader />`;

exports[`Tasks view matches snapshot when no tasks 1`] = `
.c0 {
  width: 100%;
}

.c0 td {
  border-top: 1px #f1f2f3 solid;
  font-size: 14px;
  font-weight: 400;
  padding: 8px 8px 8px 24px;
  color: #171A1C;
  vertical-align: middle;
}

.c0 tbody > tr:hover {
  background-color: #F1F2F3;
}

.c2 {
  font-family: Inter,sans-serif;
  font-size: 12px;
  font-style: normal;
  font-weight: 400;
  line-height: 16px;
  -webkit-letter-spacing: 0em;
  -moz-letter-spacing: 0em;
  -ms-letter-spacing: 0em;
  letter-spacing: 0em;
  text-align: left;
  -webkit-box-pack: start;
  -webkit-justify-content: start;
  -ms-flex-pack: start;
  justify-content: start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFFFFF;
  color: #73848C;
  cursor: default;
}

.c1 {
  padding: 4px 0 4px 24px;
  border-bottom-width: 1px;
  vertical-align: middle;
}

<table
  className="c0"
>
  <thead>
    <tr>
      <th
        className="c1"
      >
        <span
          className="c2"
        >
          ID
        </span>
      </th>
      <th
        className="c1"
      >
        <span
          className="c2"
        >
          Worker
        </span>
      </th>
      <th
        className="c1"
      >
        <span
          className="c2"
        >
          State
        </span>
      </th>
      <th
        className="c1"
      >
        <span
          className="c2"
        >
          Trace
        </span>
      </th>
      <th
        className="c1"
      >
        <span
          className="c2"
        />
      </th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td
        colSpan={10}
      >
        No tasks found
      </td>
    </tr>
  </tbody>
</table>
`;
